### ISSUE TITLE
Angular 2 http service? Should one exist and if so, what features should it have?

### ISSUE THREAD

**Issue by johnpapa** [angular/angular#2544]
Related to this https://github.com/angular/angular/pull/2360

Facts:
- We can now use any service we want without having to worry about a digest cycle
- basic XHR and HTTP are now easy from Angular

Previous $http features worth considering
- intercepting HTTP calls and extending headers, changing responses, handling errors
- easy interface
- consistent technique and all angular apps look the same (as opposed to any XHR technique you want) 

Other thoughts?

**Comment by robwormald** [comment]
It just landed in master actually: https://github.com/angular/angular/commit/21568106b114943b59ce37832d82c8fb9a63285b

Design repo - https://github.com/jeffbcross/http-design

tl;dr is its all Observable based (rather than Promises). 

@jeffbcross will correct me here if I'm wrong, but the general theory of this Http lib is that it'll make up the core of the upcoming data layer stuff, as opposed to the internal XHR implementation which is for fetching templates and the like. 

Lots of discussion on the rationale here: https://groups.google.com/forum/#!topic/angular-data-dev/tN0oGFnSEdQ - for my part, after much gnashing of teeth about losing my beloved Promises, I've more or less arrived at the conclusion that Observables don't suck.

**Comment by rolandjitsu** [comment]
:+1: for HTTP interceptors. I personally use [fetch](https://github.com/github/fetch) which works just fine with Angular since there is need to worry about digest cycles.

**Comment by jeffbcross** [comment]
The commit in #2360 was meant to provide the basis for adding many of the features you mentioned. Regarding the "should one exist" question, the goals are to provide solid conventions, simplicity, and testability. And yeah, the core does a lot to help folks use the "http of their choice" out of the box; Angular's http is opt-in.

The interceptors design is something we've started thinking about. There have been a couple of thoughts in my head about how it would be implemented (and what best practices would be considered). One strategy is similar to Angular 1's design, except less global mutable state, to add more predictability and understandability of what transformations are being applied (particularly a pain point with large teams). Another strategy would be to encourage application developers to write services to wrap `http` and apply request/response transformations inside the services. I think the service-based strategy may have some composibility costs, and could end up complicating apps more than simplifying. 

I'm doing some research in Angular 1 apps to see how interceptors are being used. It'd be great if you could provide some of the ways your team is using them in apps. Happy to jump on a hangout if that works better for you.

**Comment by rolandjitsu** [comment]
@jeffbcross [fetch](https://github.com/github/fetch) has a similar implementation to what you described.

**Comment by jeffbcross** [comment]
@rolandjitsu which part of my comment are you referring to?

**Comment by rolandjitsu** [comment]
@jeffbcross response transformations. It's up to the user to apply whatever transformations he/she needs. The response has a few methods that provide response transformations (`json`, `text` and some others), but as a user you can implement your own and apply on the promise chain (e.g. `http('some/url').then(json)` where `json` is a function that transforms the response and returns the json as a javascript object).

The transformations could be exposed as static methods on a `HTTP` class (or some more appropriate name), something like:
```js
class HTTP {
    static XML() {}
    static JSON() {}
    static Text() {}
}
```

**Comment by rolandjitsu** [comment]
@jeffbcross perhaps interceptors could be application and component specific. If a user would want an interceptor that will do something for every request/response then the user would use an application specific interceptor (and maybe injected during application bootstrap), but if the user only wants these things to happen only for a certain component than he/she would use a component interceptor (that could be configured or created with a decorator).

**Comment by jeffbcross** [comment]
@rolandjitsu actually the `response` emitted from `http` is close to the Fetch spec in how the body is accessed. See this example: https://github.com/angular/angular/blob/master/modules/examples/src/http/http_comp.ts#L28 

I'm working on a docs PR with more info that I hope to land today or tomorrow: https://github.com/angular/angular/pull/2540

**Comment by mhevery** [comment]
@jeffbcross is working on this. I am going to close this issue since there is nothing additional for us to do other than what is already in plan, and I don't want to get dawned in these kinds of issues. Feel free to continue the discussion on the closed issue.

**Comment by johnpapa** [comment]
Understood. The interceptor part seems open, based on Jeff's comment. No? If not, can we see a link to where that functionality is replaced?

**Comment by jeffbcross** [comment]
@johnpapa I'll open an issue to discuss new design in a couple of days (I have a draft in progress). In the meantime, I could use your insight on how you're using intereceptors and transformers. You can email me or comment them here.

**Comment by johnpapa** [comment]
Thanks Jeff. Feel free to move these there later. We use interceptors on http for 

- a common place to add headers to all outgoing XHR
- filter incoming responses and handle errors
- log errors for returning server error codes


We generally also create a service that handles $http so we have more control over it for our business specific rules.

**Comment by jeffbcross** [comment]
@johnpapa I opened an issue for discussion on interceptors/transformers: #2684

**Comment by danicomas** [comment]
Totally agree with your comment @johnpapa Interceptors are basically in my projects.

**Comment by gustavolira** [comment]
+1
